Original entity data resolves inverse 1-1 joins

mcurland opened the pull request:
If the source entity for an inverse (non-owning) 1-1 relationship is identified by an association then the identifying association may not be set when an inverse one-to-one association is resolved. This means that no data is available in the entity to resolve the needed column value for the join query.

Provide the original entity data for the source entity as a fallback to resolve the query conditions.

The new test will fail if the fix is removed (comment out lines 840-847 in BasicEntityPersister.php).

Fixes #11108

greg0ire commented:
Very weird… I tried locally, no issue on 2.17.x, and there were issues with your branch but… with Psalm :thinking: 
Let's close and reopen to trigger a whole run of the CI.

greg0ire commented:
Ah my bad, I forgot to fetch before trying the build on 2.17.x Now I can see I was testing your branch with Psalm 2.15.

greg0ire commented on lib/Doctrine/ORM/Cache/Persister/Entity/AbstractEntityPersister.php:
```diff
@@ -582,9 +582,9 @@ public function loadOneToManyCollection(array $assoc, $sourceEntity, PersistentC
      * {@inheritDoc}
      */
-    public function loadOneToOneEntity(array $assoc, $sourceEntity, array $identifier = [])
+    public function loadOneToOneEntity(array $assoc, $sourceEntity, array $identifier = [], array $sourceEntityData = [])
```
The class is not marked as internal. This kind of breaking change could be avoided with `func_get_args()`

mcurland replied:
I can't call that one in someone else's code base. I see some internal methods, but no internal classes. Is there anywhere you know of where this is used by something other than a UnitOfWork?

I would assume I'd get a lot of pushback from automated code checkers if I tried to slip something like this in the back door and pass more arguments than declared.

An id hash isn't available (all of the id data is still null in the entity), but the same original entity data should be available from the persister with $this->em->getUnitOfWork()->getOriginalEntityData($sourceEntity). This seems like a lot of work to do instead of just forwarding $data, but I would much rather take this approach than use func_get_args. This would also eliminate signature changes in most of the files.

Would you like me to take this approach to preserve the existing public signature?

greg0ire replied:
Yes, if by "a lot of work" you mean typing a few extra chars as opposed to a lot of work for php. I'm on my phone so can't check myself. As for internal classes, maybe they 're in another project, or just in an `Internal` namespace

mcurland replied:
The runtime hit is small, so I removed the signature change. The fix is now limited to BasicEntityPersister (plus the tests).

I'm going to tangent for an opinion (for a possible separate defect in the same code path):
The identifier is not provided by UoW for the inverse one-to-one path. This looks like a defect, and may be the source of an issue I've known about for a while. Basically, in 1-1 scenarios, if you modify an entity that was pulled as part of a 1-1 then later request a related 1-1 entity from the repository before flushing, then the entity will be replaced with fresh data from the db and your edits will be lost.

I accepted this as 'part of the library' and spit code to cache all related 1-1s when an entity is retrieved, which redirects any request for a previously fetched related 1-1 before the UoW sees it. However, entity caching is the job of the UoW, so I'd rather not keep the secondary hack cache.

I think omitting the identifier argument on the loadOneToOneEntity call in UnitOfWork.php line 2966 is the likely cause of this problem. Do you think this should be investigated? (No rush, you probably don't want to do this one from your phone).

greg0ire replied:
After reading https://www.doctrine-project.org/projects/doctrine-orm/en/2.17/reference/working-with-objects.html#entities-and-the-identity-map, my opinion is that this is indeed a bug since that behavior is not mentioned as a known issue or particular case. I think you can investigate it and that whatever you find, it could result in either a bugfix or a documentation improvement.

greg0ire commented on lib/Doctrine/ORM/Persisters/Entity/EntityPersister.php:
```diff
@@ -217,16 +217,21 @@ public function loadById(array $identifier, $entity = null);
-     * @psalm-param AssociationMapping $assoc        The association to load.
+     * @psalm-param array<string, mixed> $sourceEntityData The data used to create the sourceEntity. If the sourceEntity
+     *                                                     is identified by an association, then that association may
+     *                                                     not be initialize before this call. This original data is used
```
```suggestion
     *                                                     not be initialized before this call. This original data is used
```

greg0ire commented on tests/Doctrine/Tests/Models/OneToOneInverseSideWithAssociativeIdLoad/InverseSideIdTarget.php:
```diff
@@ -0,0 +1,33 @@
+    /**
+     * @var string
+     * @Id()
+     * @Column(type="string", length=255)
```
The length is not relevant to the issue here, is it?

```suggestion
     * @Column(type="string")
```

mcurland replied:
No, it isn't relevant here or in the OneToOneInverseSideLoad model this was directly copied from and is also the pattern in other tests I looked at. I generally have length-bounds on any column used in a key or index because dbs have limits, so I didn't think to remove it.

The more interesting question would be a different identifying type altogether (int or UUID, etc.).

The one concern I had with this fix, though, is that I wasn't sure if I should be using the raw input data. Normally this is processed through a reflField set/get value calls. I did not look into whether those ever normalize data or not. My rationale for skipping this question is that the raw data has just arrived from a SQL query and is being fed straight back into a sql query here, so should already be in a form suitable to use as a SQL parameter. Do you agree with this analysis?

greg0ire replied:
> No, it isn't relevant here or in the OneToOneInverseSideLoad model this was directly copied from and is also the pattern in other tests I looked at.

I think I would remove any distracting things that aren't strictly necessary to reproduce the issue.

> The one concern I had with this fix, though, is that I wasn't sure if I should be using the raw input data. Normally this is processed through a reflField set/get value calls. I did not look into whether those ever normalize data or not.

I don't know either, hopefully other maintainers that are more familiar with the internals of the ORM will.

> My rationale for skipping this question is that the raw data has just arrived from a SQL query and is being fed straight back into a sql query here, so should already be in a form suitable to use as a SQL parameter. Do you agree with this analysis?

I think I do, but again, I'm not sure my opinion is worth much here.

greg0ire commented on tests/Doctrine/Tests/Models/OneToOneInverseSideWithAssociativeIdLoad/InverseSide.php:
```diff
@@ -0,0 +1,34 @@
+ * @Entity()
+ * @Table(name="one_to_one_inverse_side_assoc_id_load_inverse")
+ */
+class InverseSide
```
Ok so… this entity is the inverse side of the second relationship… and the owning side of the first relationship. I think your goal was to clarify things here, but I'm afraid it achieves the opposite :sweat_smile: 

Maybe you should come up with something a more realistic example, where it is obvious which entity would own which.

mcurland replied:
This is the inverse side of the relationship the defect focuses on, which is any inverse 1-1 in an entity identified by an association. I just needed another entity for the identifier to target, so InverseSideIdTarget seemed like a reasonable name.

This was just an edit of the OneToOneInverseSideLoad model and test case, which was close to the pattern I needed. Clearly the id relationship has to be owning. The only different here is that the id column on InverseSide maps to an association and join column instead of the direct column declaration in the prior test case. If you compare the two I think it is pretty obvious what the difference is here.

The original names in the model that broke this were Action (the base type/id target here) and StateAction (the InverseSide) the other class was just State (OwningSide), which owns 1-1 Entry and Exit actions that reference StateAction. Out of context I'm not sure those names make anything any easier.

This is just a separated subtype pattern that isn't declared as such in the Doctrine ORM. It's nice to have that feature-I don't recall it being available when I first did this generated mapping in 2014 timeframe-but for the systems I'm working on this is just one of many layers generated from the same object-role model (the real breaking model produces 9k lines of Doctrine declarations sandwiched between 1400 of generated DDL and 20k lines that writes a JSON StateChange spec and reads a JSON-based query specification of the model). In other words, I very rarely interact with the Doctrine entities directly, so the final entity shape doesn't really affect me much. It doesn't matter if it is declared as a subtype or a 1-1, or if the database maps it with a separate table or combined in the same table, or even which side is owning/inverse, I just need to be able to start with any type and have the query succeed. I'm well aware this is a different perspective than using the entities as primary code artifacts, where entity aesthetics may be much more important.

greg0ire replied:
Fair enough, let's leave it as is then.

mcurland committed:
Original entity data resolves inverse 1-1 joins

If the source entity for an inverse (non-owning) 1-1 relationship is
identified by an association then the identifying association may not
be set when an inverse one-to-one association is resolved. This means
that no data is available in the entity to resolve the needed column
value for the join query.

The original entity data can be retrieved from the unit of work and
is used as a fallback to populate the query condition.

Fixes #11108

mcurland commented:
Moved from 2.17.x to 2.20.x and consolidated change sets (after we changed the approach to get the original data).

If you want this in the default branch (currently 3.2.x) instead please let me know.

Could someone please review this? IMO this was ready to go 9 months ago.

greg0ire commented:
According to https://github.com/doctrine/orm/issues/11208, since this is a bug, you should pick 2.19.x

greg0ire commented:
Rebased.

mcurland commented:
Thanks for the rebase Grégoire, even if my local is now confused : )

Now let's get this reviewed and in, please. This change is much more localized (no method signature changes) than the original proposal.

greg0ire commented:
Nothing a good old `git fetch --all && git reset --hard @{upstream}` won't fix. @derrabus @SenseException @beberlei could you please review this?

SenseException reviewed (approved):
Not sure if I understand if this is a bugfix or not, but it doesn't look wrong to put this into a current 2.x branch.

greg0ire commented:
Thanks @mcurland !
